packages/graphqlsp/src/graphql/getSchema.ts

```diff
@@ -67,147 +58,68 @@ async function saveTadaIntrospection(
   logger(`Introspection saved to path @ ${output}`);
 }
 
-export type SchemaOrigin = {
-  url: string;
-  headers: Record<string, unknown>;
-};
+export interface SchemaRef {
+  current: GraphQLSchema | null;
+  version: number;
+}
 
 export const loadSchema = (
+  // TODO: abstract info away
   info: ts.server.PluginCreateInfo,
-  schema: SchemaOrigin | string,
-  tadaOutputLocation: string | undefined,
+  origin: SchemaOrigin,
   logger: Logger
-): { current: GraphQLSchema | null; version: number } => {
-  const root =
-    resolveTypeScriptRootDir(
-      path => info.project.readFile(path),
-      info.project.getProjectName()
-    ) || path.dirname(info.project.getProjectName());
-  logger('Got root-directory to resolve schema from: ' + root);
-  const ref: {
-    current: GraphQLSchema | null;
-    version: number;
-    prev: string | null;
-  } = {
-    current: null,
-    version: 0,
-    prev: null,
-  };
-  let url: URL | undefined;
-  let config: { headers: Record<string, unknown> } | undefined;
+): SchemaRef => {
+  let loaderResult: SchemaLoaderResult | null = null;
+  const ref: SchemaRef = { current: null, version: 0 };
 
-  try {
-    if (typeof schema === 'object') {
-      url = new URL(schema.url);
-      config = { headers: schema.headers };
-    } else {
-      url = new URL(schema);
+  (async () => {
+    const rootPath =
+      (await resolveTypeScriptRootDir(info.project.getProjectName())) ||
```

Comment:
This doesn't use the internal file system host in TS anymore. But I think the way to “restore” that — or rather get rid of this entirely — is to check whether we can use an exposed API from TS to parse `tsconfig`s.